feat(state-sync): sync to the current epoch instead of the previous

marcelo-gonzalez opened the pull request:
When a node processes a block that’s the first block of epoch T, and it realizes that it will need to track shards that it doesn’t currently track in epoch T+1, it syncs state to the end of epoch T-1 and then applies chunks until it’s caught up. We want to change this so that it syncs state to epoch T instead, so that the integration of state sync/catchup and resharding will be simpler.

In this PR, this is done by keeping most of the state sync logic unchanged, but changing the “sync_hash” that’s used to identify what point in the chain we want to sync to. Before, “sync_hash” was set to the first block of an epoch, and the existing state sync logic would have us sync the state as of two chunks before this hash. So here we change the sync hash to be the hash of the first block for which at least two new chunks have been seen for each shard in its epoch. This allows us to sync state to epoch T with minimal modifications, because the old logic is still valid.

Note that this PR does not implement support for this new way of syncing for nodes that have fallen several epochs behind the chain, rather than nodes that need to catchup for an upcoming epoch. This can be done in a future PR

marcelo-gonzalez commented:
Btw, for reviewers, this PR is not quite ready to be submitted because it is only minimally tested on localnet, where I just checked that it syncs properly. You can try it with this:

```diff
diff --git a/core/primitives/src/epoch_manager.rs b/core/primitives/src/epoch_manager.rs
index 281be4399..b73b9348a 100644
--- a/core/primitives/src/epoch_manager.rs
+++ b/core/primitives/src/epoch_manager.rs
@@ -166,6 +166,8 @@ impl AllEpochConfig {
     pub fn generate_epoch_config(&self, protocol_version: ProtocolVersion) -> EpochConfig {
         let mut config = self.genesis_epoch_config.clone();
 
+        config.validator_selection_config.shuffle_shard_assignment_for_chunk_producers = true;
+
         Self::config_mocknet(&mut config, &self.chain_id);
 
         if !self.use_production_config {
```

Then if you run the transactions.py pytest, it should be able to finish after nodes sync state in the new way (you might have to comment out some asserts in that test that fail sometimes, looks unrelated but will check it)

So before submitting, I need to try this with more meaningful state and traffic/receipts, probably on forknet. Also would be good to add some integration tests, and fix whichever integration tests or pytests might have been broken by this. Also the FIXME comment in this PR needs to be fixed before I can submit this. But in any case, it should mostly be ready for review

One thing to be decided in this PR review is whether the gating via current protocol version I put in there looks okay. It feels kind of ugly to me, but it might be the easiest way to go

tayfunelmas commented on chain/chain/src/block_processing_utils.rs:
```diff
@@ -18,6 +18,17 @@ pub(crate) const MAX_PROCESSING_BLOCKS: usize = 5;
+    /// in order to sync the current epoch's state instead of last epoch's, this field being false no longer implies
+    /// that we want to apply this block during catchup, so some care is needed to ensure we start catchup at the right
+    /// point in Client::run_catchup()
     pub(crate) is_caught_up: bool,
```
is it possible to split this field into multiple fields (or enum) to differentiate these meanings? it feels like the field being false indicates both we want to apply the chunks and not apply the chunks based on other state such as sync_hash.

marcelo-gonzalez replied:
I think actually in both cases if this field is false, we don't want to apply the chunks for shards we don't currently track, and this logic should be the same: https://github.com/near/nearcore/blob/82bd46b0dba87e693a54bb4b066ccd78ab183dc1/chain/chain/src/chain.rs#L3980

I think we probably could split it, but it's a little bit tricky. Let me think about it actually... For now in this PR it is kept as is to not have to touch too many things and possibly break something. the tricky part is that right now we add the first block of the epoch to the `BlocksToCatchup` column based on this field, which is then read to see if we'll need to catch up the next block after this one as well: https://github.com/near/nearcore/blob/82bd46b0dba87e693a54bb4b066ccd78ab183dc1/chain/chain/src/chain.rs#L2288

I guess where that is called maybe we can just call `get_state_sync_info()` again, and also check if catchup is already done, but it requires some care

wacban commented on chain/chain/src/chain.rs:
```diff
@@ -3824,7 +3908,10 @@ impl Chain {
 
     /// Function to check whether we need to create a new snapshot while processing the current block
     /// Note that this functions is called as a part of block preprocesing, so the head is not updated to current block
-    fn should_make_or_delete_snapshot(&mut self) -> Result<(bool, bool), Error> {
+    fn should_make_or_delete_snapshot(
+        &mut self,
+        next_block: &BlockHeader,
+    ) -> Result<(bool, bool), Error> {
```
It seems strange that this method gets the next block as an argument and reads head at the same time. Can you pick one scheme and stick to it? If not feasible can you update the comment?

marcelo-gonzalez replied:
Yeah it is kind of strange… And note that this is done not just in the existing code in this function, but also in the caller. I would guess that you’re right and we could just use block.header().prev_hash() wherever we use the result of this head() call in here. To me seems like it def should always be the same? For now I didn’t touch any of that to avoid changing too many things, but maybe this could be made simpler. For now added a comment to the calling function

marcelo-gonzalez replied:
ok actually, I think this should be fixed... it looks like it's been this way since https://github.com/near/nearcore/pull/9090, where it's calling head() in here when it should be looking at the prev_block field. I think it would make sense to just move this out of start_process_block_impl(), because that doesnt make a lot of sense, like you pointed out in your other comment.

marcelo-gonzalez replied:
removed the block header arg from `Chain::should_make_or_delete_snapshot()` in 6963301b82f4fff65a85222447313c8a3eb7151f. Also refactored the return value a little bit to be more direct

wacban commented on chain/chain/src/chain.rs:
```diff
@@ -3836,11 +3923,26 @@ impl Chain {
         let make_snapshot = match snapshot_config.state_snapshot_type {
             // For every epoch, we snapshot if the next block would be in a different epoch
-            StateSnapshotType::EveryEpoch => is_epoch_boundary,
+            StateSnapshotType::EveryEpoch => next_block_is_sync_block,
```
nice

wacban commented on chain/chain/src/chain.rs:
```diff
@@ -4416,6 +4512,7 @@ impl Chain {
     }
 
     /// Check that sync_hash is the first block of an epoch.
+    /// TODO: allow the new way of computing the sync hash for syncing to the current epoch
```
nit: for all todos can you either add a unique name or an issue id? e.g. TODO(current_epoch_state_sync)

marcelo-gonzalez replied:
done

wacban commented on chain/chain/src/chain.rs:
```diff
@@ -4725,7 +4822,7 @@ pub struct ChunkStateWitnessMessage {
 ///     have this block as previous are added to pending
 pub struct BlocksCatchUpState {
-    /// Hash of first block of an epoch
+    /// Hash of the block where catchup will start from
     pub first_block_hash: CryptoHash,
```
Can you also update the struct comment?

marcelo-gonzalez replied:
done

wacban commented on chain/chain/src/chain_update.rs:
```diff
@@ -237,7 +237,7 @@ impl<'a> ChainUpdate<'a> {
         }
         if let Some(state_sync_info) = state_sync_info {
-            self.chain_store_update.add_state_sync_info(state_sync_info);
+            self.chain_store_update.add_state_sync_info(*block.hash(), state_sync_info);
```
Why not put the block hash inside of the state sync info?

marcelo-gonzalez replied:
See the comment above about avoiding the DB migration

marcelo-gonzalez replied:
reverted to the old code where we dont pass this block hash anymore

wacban commented on chain/chain/src/store_validator.rs:
```diff
@@ -249,13 +249,6 @@ impl StoreValidator {
                     let block_hash = CryptoHash::try_from(key_ref)?;
                     let state_sync_info = StateSyncInfo::try_from_slice(value_ref)?;
-                    // StateSyncInfo is valid
-                    self.check(
```
Why did you remove this check?

marcelo-gonzalez replied:
Also see the comment above about avoiding the DB migration

wacban commented on chain/client/src/client.rs:
```diff
@@ -111,6 +111,12 @@ pub enum AdvProduceBlocksMode {
     OnlyValid,
 }
 
+pub struct CatchupState {
+    pub state_sync: StateSync,
+    pub state_downloads: HashMap<u64, ShardSyncDownload>,
+    pub catchup: BlocksCatchUpState,
+}
+
```
add comments please

marcelo-gonzalez replied:
done

wacban commented on chain/client/src/client.rs:
```diff
@@ -147,10 +153,12 @@ pub struct Client {
     /// Approvals for which we do not have the block yet
     pub pending_approvals:
         lru::LruCache<ApprovalInner, HashMap<AccountId, (Approval, ApprovalType)>>,
+    /// A mapping from the first block of an epoch that we know needs to be state synced for some shards
+    /// to a tracker that will find an appropriate sync_hash for state sync
+    catchup_tracker: HashMap<CryptoHash, NewChunkTracker>,
```
mini nit: Intuitively it would make sense to track state sync per epoch (epoch_id). Would it make sense in practice?

marcelo-gonzalez replied:
yup, agreed

wacban replied:
<brainstorm> Perhaps we could even change the sync hash to state sync id that is the epoch id. I'm not sure if it makes sense, how much work would it take and if it's worth it though </brainstorm>

wacban commented on chain/client/src/client.rs:
```diff
@@ -2448,6 +2542,23 @@ impl Client {
+    fn set_state_sync_hash<'a>(
+        mut update: ChainStoreUpdate<'a>,
+        epoch_first_block: CryptoHash,
+        state_sync_info: &mut StateSyncInfo,
```
Why does this need to be mutable?

marcelo-gonzalez replied:
i did this just as a helper to set the in-memory and on-disk values at the same time, but actually let's just remove it. I think I thought I would need it in more than one place when I put it there

wacban commented on chain/client/src/client.rs:
```diff
@@ -2448,6 +2542,23 @@ impl Client {
     }
 
+    fn set_state_sync_hash<'a>(
+        mut update: ChainStoreUpdate<'a>,
```
Can this be a method on ChainStoreUpdate?

marcelo-gonzalez replied:
just removed it

wacban reviewed (commented):
I didn't manage to get through everything but I left some comments to keep you busy for now ;) Please have a look and let's reconvene. 

It may be easier to review and merge this if you can split it down into smaller PRs. Not sure if it makes sense but something to consider. 

I still need to catch up on the reading to understand why we need to have two new chunks in every shard. It would make it way easier with one so I want to make sure. I'll have another look at your proposal doc and come back. 

Once this is done can you run nayduck and make sure that all the tests are still passing? 

I think it would be best to add to nightly first and stabilize only after DSS is released.

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -781,8 +781,15 @@ impl Chain {
+            let protocol_version =
+                self.epoch_manager.get_epoch_protocol_version(block.header().epoch_id())?;
+            let sync_hash =
+                if protocol_version < 72 { *block.header().hash() } else { CryptoHash::default() };
```
made it a nightly feature

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -781,8 +781,15 @@ impl Chain {
+            let protocol_version =
+                self.epoch_manager.get_epoch_protocol_version(block.header().epoch_id())?;
+            let sync_hash =
+                if protocol_version < 72 { *block.header().hash() } else { CryptoHash::default() };
```
I agree that magic values are ugly in general, but here it’s done to avoid the need for a database migration. 

So before this change, this field is always set to the hash of the first block of the epoch (which is the sync_hash we use for state sync). And this is also used as the key we look up this value with. But this `sync_hash` field in the `StateSyncInfo` struct is actually only ever checked once, where we just [check if it’s equal to the key that looked up that struct](https://github.com/near/nearcore/blob/70acea66bac9e5eeee64c0f417112f762e2a2b55/chain/client/src/client.rs#L2467).

So this `sync_hash` field is actually just redundant data and is not used meaningfully, which is why I took the opportunity to change the way we set it, because it can be done with logic that’s consistent before/after the binary upgrade. We just set it to zero after the protocol change, and if we read it and that field is set, then that’s the sync hash we want to use, otherwise, we’re still looking for the sync hash.

So all this is just to avoid a DB migration. I felt that if we could do so, it’s not so bad to accept slightly ugly code with a magic value of zeros, because a DB migration is a sort of nonzero development/UX churn point. But wdyt? I could of course just do this the right way and add a DB migration if we don’t think it’s such a big deal

wacban replied:
Thanks for sharing context, I didn't realize it would necessitate a db migration. 

Still I think state sync being in a pretty bad place as it is I wouldn't want to add more tech debt to it. I don't think a db migration is a big deal even if it takes a bit extra effort. I would expect the state sync info to be minimal in size so the migration should be really quick too. Can you try and see how bad is it?

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -1797,7 +1804,7 @@ impl Chain {
         let (apply_chunk_work, block_preprocess_info) = preprocess_res;
 
         // 2) Start creating snapshot if needed.
-        if let Err(err) = self.process_snapshot() {
+        if let Err(err) = self.process_snapshot(block.header()) {
             tracing::error!(target: "state_snapshot", ?err, "Failed to make a state snapshot");
         }
```
This is a v good question, I am not sure… I think it should be pretty possible to just decouple this, since it’s not needed in any block/chunk processing. My guess is it was just easy to express it as a callback that gets called every time we update the HEAD, in a way that you could sort of imagine the indexer being rewritten to do (instead of polling every so often for a new block)

wacban replied:
I'm fine having this check to happen on every block, it's a common pattern. It just seems strange that it happens before the block is applied. If the block turns out to be invalid the node may do an "invalid" snapshot for some definition of invalid.

marcelo-gonzalez replied:
wait yea that is a good point.. and I feel like not only should we wait until it's applied, but maybe we should wait until it's final too

wacban replied:
I know we discussed it but I don't remember the final conclusion - can you remind me please?

marcelo-gonzalez replied:
Oh I think it's just a little bit tricky where we have to make sure flat storage is still available when we make the snapshot, and that the flat storage head hasn't moved past the point where we want to take the snapshot. I'm not 100% sure what the guarantees are on that, but keeping this here at least preserves the old behavior, and in another PR if it makes sense we can move this. But worth noting that in one of the recent commits addressing your comment about not passing the block header to `Chain::should_make_or_delete_snapshot()`, we now take a snapshot for the current epoch sync hash one block later than before, and it still works

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -2407,6 +2414,84 @@ impl Chain {
         )
     }
 
+    fn get_epoch_start_sync_hash_impl(
```
Yeah I agree, although here I’m actually just moving the existing code from [here](https://github.com/near/nearcore/blob/70acea66bac9e5eeee64c0f417112f762e2a2b55/chain/client/src/sync/state.rs#L464). For now I just left it as-is without trying to optimize it any further, but maybe in another PR we should do something like that

wacban replied:
Cool, it's fine if you want to do it in a follow up, can you just add a TODO in there?

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -2407,6 +2414,84 @@ impl Chain {
+    }
+
+    // Returns the first block for which at least two new chunks have been produced for every shard in the epoch
+    pub fn get_current_epoch_sync_hash(
```
Hmm, yea like I said above I moved get_epoch_start_sync_hash() out of the state sync code, but maybe we could move them both back over there? The reason I moved it to chain.rs is that previously, chain.rs was not referencing the state sync code to [decide whether it wants to make a new snapshot](https://github.com/near/nearcore/blob/70acea66bac9e5eeee64c0f417112f762e2a2b55/chain/chain/src/chain.rs#L3922). To do that, it could just check if the block was the first of the epoch. But really what we’re checking is whether the next block will be the “sync_hash” block. It’s just that those two happened to be the same before. But now that there’s some more logic involved in deciding that, I moved it here to not make chain.rs depend on any state sync files. But actually I think after I do the FIXME in here (that says we should just not call this inefficient code in this spot anyway), I think we’ll be able to move get_epoch_start_sync_hash() back to where it was before, because here we won’t need to call that in deciding if the next block is the sync block

wacban replied:
Just to be clear, I'm not suggesting to do it in this PR. Just generally state sync could use some refactoring. If you see any potential improvements like those it would be awesome to keep track of it and implement some.

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -2407,6 +2414,84 @@ impl Chain {
+    // Returns the first block for which at least two new chunks have been produced for every shard in the epoch
+    pub fn get_current_epoch_sync_hash(
+        &self,
+        sync_header: &BlockHeader,
```
This part is because actually if you just pass a sync_hash to it, it will fail in get_block_header() because this function is called in start_process_block_impl() on a block header that has not yet been saved. So if you call get_epoch_start_sync_hash  on this block, it will fail. But we do actually have the block header from the caller, so we can just pass it as an argument even though it’s not saved to the DB yet

wacban replied:
Ah I think it's just the naming that confused me, can you clarify what is the sync header and / or rename it appropriately?

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -2407,6 +2414,84 @@ impl Chain {
+        let mut num_new_chunks: HashMap<_, _> =
+            shard_ids.iter().map(|shard_id| (*shard_id, 0)).collect();
+
+        loop {
```
Yea agreed about not iterating over the blocks, was going to address exactly that in the FIXME comment in here. Actually yeah I think what you suggest there should work, with some care to make sure we get the same point in the chain as other nodes that are state sync requesters looking for the right sync_hash (but I think with what you suggest it should work out to be the same)

marcelo-gonzalez replied:
will do later in another PR, added TODO here

marcelo-gonzalez commented:
> It may be easier to review and merge this if you can split it down into smaller PRs. Not sure if it makes sense but something to consider.

Actually here I tried to make the individual commits inside this PR somewhat well contained on their own, so hopefully those are easier to look at?

marcelo-gonzalez commented:
So added a DB migration to store a new cleaner version of the `StateSyncInfo` struct. I added a `state_sync_version` field to it currently just set to 0, but I think it might be helpful if we want to change things about the meaning of the sync hash in the future. We really should only need to wait for one new chunk per shard, and the reason we wait for two in this PR is because it's easier to implement a first version doing that. But if we want to make that change then we'll need to indicate what we mean by the sync hash somehow.

Also see this commit heheh... https://github.com/near/nearcore/pull/12102/commits/78d6a756d5dbb5347b1d57c78cb1170dd356f832 . I removed the part that says "before submitting", because I think it's probably okay to submit it as-is for now, since it's mostly going to be fine as long as we make sure to fix it before this makes it into a release. Fixing that seems like a pretty well-scoped individual PR that would be a little easier to review on its own

let me know what ppl think

marcelo-gonzalez commented:
also there's a test here to make sure the migration from the previous `StateSyncInfo` to the new works okay: https://github.com/marcelo-gonzalez/nearcore/blob/catchup-migration-test/pytest/tests/sanity/state_sync_info_migration.py

don't think that's really worth adding here though, but good to sanity check it once

marcelo-gonzalez commented:
aghh, just noticed the nightly test is failing... will fix it tmr

wacban reviewed (commented):
some partial review, to be continued

wacban commented on chain/chain/src/chain.rs:
```diff
@@ -782,8 +782,17 @@ impl Chain {
             let state_sync_info = StateSyncInfo {
-                epoch_tail_hash: *block.header().hash(),
+                state_sync_version: 0,
+                epoch_first_block: *block.header().hash(),
```
It's non obvious that the block is guaranteed to be the first in this epoch - can you clarify that somehow?

marcelo-gonzalez replied:
Agreed, p non obvious. This function is only called when the block is the first of an epoch, and this implicit assumption here that it’s the first of an epoch existed even before this PR. Renamed the function argument and added a comment

wacban commented on nearcore/src/state_sync.rs:
```diff
@@ -662,12 +668,21 @@ fn get_latest_epoch(
+    // This is not the way protocol features are normally gated, but this is not actually
+    // a protocol feature/change. The protocol version is just an easy way to coordinate
+    // among nodes for now
+    let sync_hash =
+        if checked_feature!("stable", StateSyncHashUpdate, epoch_info.protocol_version()) {
+            chain.get_current_epoch_sync_hash(&final_block_header)?
+        } else {
+            Some(chain.get_epoch_start_sync_hash(final_hash)?)
+        };
```
I like this pattern a lot. I would expect it to appear in all the places where we need to figure out the right sync hash. Would it make sense to refactor this to a method? Ideally the feature check should happen in just one place and then the logic should depend on the type of state sync that is encoded in the structs. 

Also the naming is a bit inconsistent - how about the following? Would be nice to make those private too. The general idea is to pick a name for each type of state sync and use those names everywhere. I think you're doing it already with `current epoch state sync` which is nice. 
* get_current_epoch_sync_hash 
* get_previous_epoch_sync_hash

If it so happens that the method is reused for getting the sync hash and the first block of an epoch - let's just add a new method for sync hash and call the one for getting epoch's first block.

marcelo-gonzalez replied:
yea good idea

wacban commented on core/store/src/migrations.rs:
```diff
@@ -358,3 +360,39 @@ pub fn migrate_39_to_40(store: &Store) -> anyhow::Result<()> {
+
+    let mut update = store.store_update();
+
+    for row in store.iter_prefix_ser::<LegacyStateSyncInfo>(DBCol::StateDlInfos, &[]) {
```
mini nit: it's a bit funny to call `iter_prefix` and use empty prefix - if it's simple enough I would suggest implementing `iter_ser` (assuming it's missing).

marcelo-gonzalez replied:
implemented and also updated one other caller in this migrations.rs file

wacban commented on core/store/src/migrations.rs:
```diff
@@ -358,3 +360,39 @@ pub fn migrate_39_to_40(store: &Store) -> anyhow::Result<()> {
+            sync_hash: Some(sync_hash),
+            shards,
+        };
+        update
+            .set_ser(DBCol::StateDlInfos, &key, &new_info)
+            .context("failed writing to StateDlInfos")?;
```
Sanity check - is there any potential for this update to OOM?

wacban replied:
I guess I just don't know how does cleanup work for those - if you happen to know please share :)

marcelo-gonzalez replied:
I think in practice there's really only going to be one or zero keys in the `StateDlInfos` column, since the key is the hash of the first block of an epoch, and it's cleaned up after catchup is done, so really we only see one iteration of this loop in practice

wacban commented on chain/client/src/client.rs:
```diff
@@ -2463,42 +2562,97 @@ impl Client {
-        for (sync_hash, state_sync_info) in self.chain.chain_store().iterate_state_sync_infos()? {
-            assert_eq!(sync_hash, state_sync_info.epoch_tail_hash);
-            let network_adapter = self.network_adapter.clone();
+        for (epoch_first_block, mut state_sync_info) in
```
this method is too large - can you split it down into smaller methods please?

marcelo-gonzalez replied:
extracted a get_catchup_sync_hash() function in f96f76203a4c802466f4c14bfb8bdabc5ca4635b

wacban commented on chain/client/src/client.rs:
```diff
@@ -229,6 +242,91 @@ pub struct ProduceChunkResult {
+        Ok(done)
+    }
+
+    fn find_sync_hash(&mut self, chain: &Chain) -> Result<Option<CryptoHash>, Error> {
```
I think I saw another method with the same name. Is there any chance to merge those or make one a thin wrapper of the other?

marcelo-gonzalez replied:
^ like in the above comment, I plan to delete `get_current_epoch_sync_hash` in another PR, so hopefully this will all just be in one place. If you think it should be done in this PR though, let me know

wacban reviewed (commented):
A few more comments, mostly stylistic but can you address those before the next round please? Feel free to "resolve" the comments that are done. Please ping me when ready.

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -783,8 +783,18 @@ impl Chain {
+            // among nodes for now
+            let protocol_version =
+                self.epoch_manager.get_epoch_protocol_version(block.header().epoch_id())?;
+            let sync_hash = if checked_feature!("stable", StateSyncHashUpdate, protocol_version) {
```
Yeah I like that better

marcelo-gonzalez commented on core/primitives/src/sharding.rs:
```diff
@@ -56,14 +56,29 @@ impl From<CryptoHash> for ChunkHash {
+    /// is that when syncing to the current epoch's state, we currently wait for two new chunks in each shard, but
+    /// with some changes to the meaning of the "sync_hash", we should only need to wait for one. So this is included
+    /// in order to allow for this change in the future without needing another database migration.
+    pub state_sync_version: u32,
```
Yeah I think that’s prob better, just in case there’s something about the struct we want to change without needing a db upgrade. Updated, and also tested again with https://github.com/marcelo-gonzalez/nearcore/blob/catchup-migration-test/pytest/tests/sanity/state_sync_info_migration.py

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -2415,6 +2424,89 @@ impl Chain {
+        loop {
+            header = if header.hash() == next_header.prev_hash() {
+                next_header.clone()
+            } else {
```
Yeah that was the reason why. But now after addressing your other comment about only relying on head() in should_make_or_delete_snapshot(), this goes away

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -3843,11 +3938,27 @@ impl Chain {
+                is_epoch_boundary
+            } else {
+                // FIXME: this needs to be fixed. can't be iterating over the whole chain inside of preprocess
+                // block like that if there are many missed chunks
```
Yup, will do

marcelo-gonzalez commented on chain/client/src/client.rs:
```diff
@@ -111,6 +111,17 @@ pub enum AdvProduceBlocksMode {
+    /// Manages downloading the state.
+    pub state_sync: StateSync,
+    /// Keeps track of state downloads, and gets passed to `state_sync`.
+    pub state_downloads: HashMap<u64, ShardSyncDownload>,
```
done

marcelo-gonzalez commented on chain/client/src/client.rs:
```diff
@@ -229,6 +242,91 @@ pub struct ProduceChunkResult {
+/// This whole thing could be replaced with a much simpler function that just computes the numbef of new chunks
+/// in each shard from scratch every time we call it, but in the unlikely and unfortunate case where a shard
+/// hasn't had any chunks for a very long time, it would end up being a nontrivial inefficiency to do that
+/// every time run_catchup() is called
```
So now that I think about your suggestion again, I think it would not quite result in the same sync hash (if I understnad your suggestion correctly, we just check `height_included` for each chunk and see if it's new or not? The issue I think is that we actually want two blocks with new chunks in every shard after the first block of the epoch (so usually the third block of the epoch). So unless I'm missing something, I think there has to be some sort of iteration. I think in another PR that fixes the FIXME in this one, we can figure something better out to reduce the number of places where this calculation is done (maybe for example saving this info for each block in the DB so we can just look it up, and when we apply a new block, we look up the info for its parent?). But if you have other ideas, I'm open to them

marcelo-gonzalez commented on chain/client/src/client.rs:
```diff
@@ -229,6 +242,91 @@ pub struct ProduceChunkResult {
+        }
+    }
+
+    fn record_new_chunks(&mut self, header: &BlockHeader) -> Result<bool, Error> {
```
Yeah, I think actually in another PR I want to just delete `get_current_epoch_sync_hash` altogether (like I said in the above comment), so after that, this logic should only be in one place

marcelo-gonzalez commented:
@wacban updated with fixed tests and more commits addressing comments, PTAL. There's a decent amount changed from last time, but I tried to keep the commits small, so I think it will be much easier to review if you just look at the individual commits from last time you looked. Also haven't merged it with master for now because I heard there were some state sync problems on master, but will do that later

wacban reviewed (commented):
A few more comments, please ping me once updated with the recent master changes.

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -782,16 +792,27 @@ impl Chain {
+            let protocol_version = self
+                .epoch_manager
+                .get_epoch_protocol_version(epoch_first_block.header().epoch_id())?;
+            let sync_hash = if ProtocolFeature::StateSyncHashUpdate.enabled(protocol_version) {
+                None
+            } else {
+                Some(*epoch_first_block.header().hash())
+            };
```
Ah no here actually it's not possible because we're inside `preprocess_block()`, so `epoch_first_block` has not yet been processed and saved as HEAD. So then if you try calling `get_sync_hash(epoch_first_block.header().hash())`, you'll get a `DBNotFoundErr` when looking up that block's hash.

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -3783,25 +3865,35 @@ impl Chain {
-        }
+        match snapshot_action {
+            SnapshotAction::MakeSnapshot(prev_hash) => {
+                let block = self.get_block(&prev_hash)?;
```
Yeah, that's right. the prev prev hash of the sync hash is what should be the prev hash below

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -3831,32 +3923,64 @@ impl Chain {
-        let make_snapshot = match snapshot_config.state_snapshot_type {
-            // For every epoch, we snapshot if the next block would be in a different epoch
-            StateSnapshotType::EveryEpoch => is_epoch_boundary,
+        match snapshot_config.state_snapshot_type {
+            // For every epoch, we snapshot if the next block is the state sync "sync_hash" block
```
Here I didn't do that because after removing the block header argument to this function, now the current block being preprocessed is not saved as the HEAD. So if that block is the "sync_hash" block, we won't be able to detect it until the next block. That's why below when we call `get_current_epoch_sync_hash()`, and it's equal to `head.last_block_hash`, we return `head.prev_block_hash`. Whereas in the prev epoch sync case, we actually do have the luxury of checking whether the next block is they sync block even without passing the block header being preprocessed to this function, because we can just use `epoch_manager.is_next_block_epoch_start()`.

So by not using `get_sync_hash()` and instead writing out each case explicitly, we can preserve the old behavior for the old state sync method where we take a snapshot when preprocessing the sync hash. It's probably not a huge deal but it means less is different, and so not as much of a chance of some unforseen bug. WDYT though? We could just snapshot one block later in both cases, but I wanted to change as little as possible

marcelo-gonzalez replied:
Ok I'll just leave it for now since we're probably going to need to change the snapshot logic anyway after the changes to flat storage during resharding

marcelo-gonzalez commented on core/primitives/src/sharding.rs:
```diff
@@ -56,18 +56,47 @@ impl From<CryptoHash> for ChunkHash {
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize)]
+pub enum StateSyncInfo {
+    V0(StateSyncInfoV0),
+}
```
sounds good, done

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -2415,6 +2437,67 @@ impl Chain {
     }
 
+    // TODO(current_epoch_state_sync): move state sync related code to state sync files
+    pub fn get_sync_hash(&self, block_hash: &CryptoHash) -> Result<Option<CryptoHash>, Error> {
```
done

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -2415,6 +2437,67 @@ impl Chain {
+                            header.hash()
+                        )));
+                    }
+                };
```
done

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -3831,32 +3923,64 @@ impl Chain {
+                    }
+                } else {
+                    // FIXME: this needs to be fixed. can't be iterating over the whole chain inside of preprocess
+                    // block like that if there are many missed chunks
```
done

marcelo-gonzalez commented on chain/chain/src/chain.rs:
```diff
@@ -3831,32 +3923,64 @@ impl Chain {
+                    // block like that if there are many missed chunks
+                    match self.get_current_epoch_sync_hash(&head.last_block_hash)? {
+                        Some(sync_hash) => {
+                            if sync_hash == head.last_block_hash {
```
done

wacban reviewed (approved):
LGTM, thanks! 

I think we should fix the duplication of code for finding the sync hash but it can be done separately.

marcelo-gonzalez commented on chain/client/src/client.rs:
```diff
@@ -139,9 +151,12 @@ pub struct Client {
         lru::LruCache<ApprovalInner, HashMap<AccountId, (Approval, ApprovalType)>>,
+    /// A mapping from an epoch that we know needs to be state synced for some shards
+    /// to a tracker that will find an appropriate sync_hash for state sync to that epoch
+    catchup_tracker: HashMap<EpochId, NewChunkTracker>,
```
You know, that is a good question... It seems this mapping from epoch ID to catchup status has been around since state sync was first implemented, and looking at the code it doesn't seem to me like it should ever have more than one... When we add one, it's because we see a block in a new epoch T, and we know we're going to be a chunk producer for T+1 in a new shard. The epoch ID for T+1  is the hash of the last block in epoch T-1, so if there's a fork at the beginning of the epoch, it will still have the same epoch ID. So that should mean that for any given epoch height, we only have one of these. And then if we ask if it's possible to have epoch T and T+1 in there at the same time, we can look at the fact that we remove the epoch info for epoch T in `finish_catchup_blocks()` when we call `remove_state_sync_info()`. If that hasn't happened by the time we get to the first block of epoch T+1, we will not add another catchup state keyed by the epoch ID of epoch T+1, because we'll [call that block an orphan](https://github.com/near/nearcore/blob/bf28607aa18b59e4441b61e9ffb73b4cdc18c1fc/chain/chain/src/chain.rs#L2344) until [we remove the last block of epoch T](https://github.com/near/nearcore/blob/bf28607aa18b59e4441b61e9ffb73b4cdc18c1fc/chain/chain/src/chain.rs#L3284) from the `BlocksToCatchup` column in `finish_catchup_blocks()`  .

So idk how it is even possible to have two, but maybe I'm missing something and that logic was put there for a reason?

But now that I think about it, what happens if we apply the first block of a new epoch and then save `StateSyncInfo` for it, which on the DB is keyed by the hash of that first block, and then that block doesn't end up on the main chain, because there's a fork off the last block of epoch T-1? Is there some reason that won't happen for the first block of an epoch? I'm not sure if there is, but if so, then there's an implicit assumption here that at least should have gotten a comment explaining, and if not, there's a potential bug here. I think in general maybe we should only be working with final blocks when doing all this state sync stuff. If it is a bug, then for now it's a pretty unlikely one I guess, but it's worth investigating more. Shouldn't be too too hard to try causing a fork at the first block of an epoch in a testloop test and seeing what happens

marcelo-gonzalez replied:
ehh it's possible that last part is not a bug actually, since there's care put into only removing the particular sync hash from the `BlocksToCatchup` [here](https://github.com/near/nearcore/blob/bf28607aa18b59e4441b61e9ffb73b4cdc18c1fc/chain/chain/src/chain.rs#L3275), but it's a bit confusing and worth sanity checking

marcelo-gonzalez commented:
fyi, @wacban added a couple more commits to fix tests. Note that in 9991c89c9295874d853de15b48abff1a30d4e072 I'm returning early from get_sync_hash() for the genesis block and in 99f98a5f64edabb9dfb4bc461f95da93c4569d5e just not even checking what happens with the genesis block in tests because I don't think we ever want to be using the genesis block here in any case. `ClientActorInner::find_sync_hash()` even has an assertion against it. Let me know if you have objections though :)
